File: server/src/models/rbac.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct Permission {
    pub id: Uuid,
    pub permission_string: String,
    pub description: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct AppRole {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    /// `global` or `course` (course-scoped roles are documented for per-course assignment).
    pub scope: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RoleWithPermissions {
    #[serde(flatten)]
    pub role: AppRole,
    pub permissions: Vec<Permission>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatePermissionRequest {
    pub permission_string: String,
    pub description: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatchPermissionRequest {
    pub description: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRoleRequest {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default = "default_role_scope")]
    pub scope: String,
}

fn default_role_scope() -> String {
    "global".to_string()
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatchRoleRequest {
    pub name: String,
    pub description: String,
    pub scope: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetRolePermissionsRequest {
    pub permission_ids: Vec<Uuid>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PermissionsListResponse {
    pub permissions: Vec<Permission>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RolesListResponse {
    pub roles: Vec<RoleWithPermissions>,
}

#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct UserBrief {
    pub id: Uuid,
    pub email: String,
    pub display_name: Option<String>,
    pub sid: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RoleUsersResponse {
    pub users: Vec<UserBrief>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CourseScopedRolesResponse {
    pub roles: Vec<AppRole>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddRoleUserRequest {
    pub user_id: Uuid,
}

File: server-new/internal/models/rbac/rbac.go
// Package rbac contains JSON DTOs for the settings roles/permissions API (Rust server/src/models/rbac.rs).
package rbac

import (
	"time"

	"github.com/google/uuid"
)

// Permission is a row in "user".permissions.
type Permission struct {
	ID                 uuid.UUID `json:"id"`
	PermissionString  string    `json:"permissionString"`
	Description       string    `json:"description"`
	CreatedAt         time.Time `json:"createdAt"`
}

// AppRole is a row in "user".app_roles.
type AppRole struct {
	ID          uuid.UUID `json:"id"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Scope       string    `json:"scope"`
	CreatedAt   time.Time `json:"createdAt"`
}

// RoleWithPermissions is the app role with its permission rows (JSON shape uses flattened role fields).
type RoleWithPermissions struct {
	AppRole
	Permissions []Permission `json:"permissions"`
}

// UserBrief is a minimal user for role membership lists.
type UserBrief struct {
	ID          uuid.UUID `json:"id"`
	Email       string    `json:"email"`
	DisplayName *string   `json:"displayName"`
	Sid         *string   `json:"sid"`
}

// CreatePermissionRequest is the body for POST /api/v1/settings/permissions.
type CreatePermissionRequest struct {
	PermissionString string `json:"permissionString"`
	Description      string `json:"description"`
}

// PatchPermissionRequest is the body for PATCH /api/v1/settings/permissions/{id}.
type PatchPermissionRequest struct {
	Description string `json:"description"`
}

// CreateRoleRequest is the body for POST /api/v1/settings/roles.
type CreateRoleRequest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Scope       string `json:"scope"`
}

// PatchRoleRequest is the body for PATCH /api/v1/settings/roles/{id}.
type PatchRoleRequest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Scope       string `json:"scope"`
}

// SetRolePermissionsRequest is the body for PUT /api/v1/settings/roles/{id}/permissions.
type SetRolePermissionsRequest struct {
	PermissionIDs []uuid.UUID `json:"permissionIds"`
}

// AddRoleUserRequest is the body for POST /api/v1/settings/roles/{id}/users.
type AddRoleUserRequest struct {
	UserID uuid.UUID `json:"userId"`
}

// PermissionsListResponse is GET /api/v1/settings/permissions.
type PermissionsListResponse struct {
	Permissions []Permission `json:"permissions"`
}

// RolesListResponse is GET /api/v1/settings/roles.
type RolesListResponse struct {
	Roles []RoleWithPermissions `json:"roles"`
}

// RoleUsersResponse is user lists under a role.
type RoleUsersResponse struct {
	Users []UserBrief `json:"users"`
}
